lsp: correct float placement when using borders

mhinz opened the pull request:
Because of the additional 2 lines when using borders, we need to subtract 2 from
the anchor position, when opening a float in the lower half of the window.

mhinz committed:
lsp: make sure borders are always inside of the screen

Add a helper function to get the additional width a border adds to a float to
make sure that the border is always inside the visible screen.

mjlbach commented:
Does this depend on the size of the borders? Not sure if we can add multiple border chars currently

mhinz commented:
I think there can only be a one-cell border, but I'll double-check this.

And yes, there's definitely something wrong with the width calculation. Even when using the default border on the left and right, I don't see the padding on the right side, because `_make_floating_popupsize()` returns a totally wrong width. I'll look into that and add the fix to this PR.

mhinz committed:
lsp: correct float placement when using borders

Because borders add up to 2 to the height of a float, we need to subtract that
from the anchor position, when opening a float in the lower half of the window.

mhinz commented:
That border handling requires more code than I initially assumed, but I see no other way as long as we allow borders.

mjlbach commented:
>  but I see no other way as long as we allow borders.

Do you think we should remove allowing customizing borders and have people specify their own padding values? (serious question).

The only reason I enabled borders (which I initially did not style, as seen with the defaults) was to eliminate the padding issues with the old trim_and_pad function in fancy floating markdown which led to syntax highlighting errors.

Otherwise the PR LGTM, I probably would have made the call to get_border_size pure and not access the default border function (instead pass in the true border).

mhinz commented:
> Do you think we should remove allowing customizing borders and have people specify their own padding values? (serious question).

Well, personally I do like a single border and just padding wouldn't do it for me. :) I just think the float is easier to see that way.

> The only reason I enabled borders (which I initially did not style, as seen with the defaults) was to eliminate the padding issues with the old trim_and_pad function in fancy floating markdown which led to syntax highlighting errors.

Oh, so it would still be an issue if someone passed a custom border with all cells set to `''`, eh? I can look into that highlighting issue.

> Otherwise the PR LGTM, I probably would have made the call to get_border_size pure and not access the default border function (instead pass in the true border).

At the time the function was called, `opts` wouldn't contain the default border yet. So, I could have assumed `if not opts.border`, then I know the default border will be set later and I can set border width to 2 and height to 0. But then it would break if the default should ever change again etc.

Personally I would be okay with just _always_ using the full screen width instead. That would save a lot of width and wrapping calculations and make borders obsolete as well. But I guess that would be a tad too extreme for some users (and doesn't allow for pretty screenshots!) :]
